### DISCUSSION TITLE
@Configuration과 @Component의 차이

### DISCUSSION THREAD

**Discussion by Kimprodp** [2024-woowacourse-study/level-interview#128] [Spring Core]
component scan

**Comment by alstn113** [comment]
### @Component

- **용도**: 개발자가 직접 작성한 클래스를 스프링 빈으로 등록하고자 할 때 사용합니다.
- **특징**: 
  - 자동 스캔을 통해 스프링 컨테이너에 빈으로 등록됩니다.
  - `@Controller`, `@Service`, `@Repository` 등의 애노테이션이 `@Component`를 상속받아 특정 역할을 가진 빈을 등록할 때 사용됩니다.

### @Configuration

- **용도**: 외부 라이브러리나 개발자가 직접 제어할 수 없는 클래스를 빈으로 등록하고자 할 때 사용합니다.
- **특징**:
  - 한 개 이상의 `@Bean` 메서드를 제공하는 클래스를 정의할 때 사용합니다.
  - 클래스 내부에 정의된 `@Bean` 메서드를 통해 스프링 빈을 등록하고 관리합니다.
  - `@Configuration` 클래스 자체도 스프링 빈으로 등록되며, 내부의 `@Bean` 메서드들은 싱글톤으로 관리됩니다.

**Comment by 3Juhwan** [comment]
빈을 등록하는 방법은 자동 등록과 수동 등록이 있습니다. `@Component`는 자동 등록에,  `@Configuration`과 `@Bean`은 수동 등록에 사용합니다. 

**`@Component`**

- 클래스 레벨에 선언할 수 있습니다. 
- 컴포넌트 스캔을 통해 빈으로 자동 등록됩니다. 
- 개발자가 제어할 수 있는 클래스에만 사용할 수 있습니다. 가령, ObjectMapper는 `@Component`을 선언할 수 없어서 인스턴스를 만들고 `@Bean`을 선언해야 합니다. 

**`@Bean`**

- 메서드 레벨에 선언할 수 있습니다. 
- `@Configuration`가 선언된 클래스 내에서 빈으로 수동 등록됩니다. 

**`@Configuration`**

- `@Configuration` 클래스를 사용하면 해당 클래스가 Spring IoC 컨테이너에 의해 관리되는 설정 클래스입니다.
- `@Bean` 어노테이션이 붙은 메서드는 메서드의 반환 값을 스프링 컨테이너의 빈으로 등록합니다. 
- `@Component`를 포함하고 있어 스프링 컨테이너에 의해 스캔되고 빈으로 등록됩니다.
